File: 프로그래머스/1/159994. 카드 뭉치/카드 뭉치.cs
using System;

public class Solution 
{
    public string solution(string[] cards1, string[] cards2, string[] goal) 
    {
        int i = 0;  // cards1의 다음에 뽑을 위치
        int j = 0;  // cards2의 다음에 뽑을 위치
        
        foreach (string word in goal)
        {
            if (i < cards1.Length && cards1[i] == word)
            {
                i++;
            }
            else if (j < cards2.Length && cards2[j] == word)
            {
                j++;
            }
            else
            {
                return "No";  // 어느 쪽 맨 앞과도 안 맞으면 완성 불가능
            }
        }
        
        return "Yes";
    }
}

File: 프로그래머스/1/159994. 카드 뭉치/카드 뭉치.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

string solution(vector<string> cards1, vector<string> cards2, vector<string> goal) 
{
    map<string, int> cardsMap1;
    map<string, int> cardsMap2;
    
    for(int i=0; i<cards1.size(); i++) { cardsMap1[cards1[i]] = i; }
    for(int i=0; i<cards2.size(); i++) { cardsMap2[cards2[i]] = i; }
    
    vector<int> idxVec1;
    vector<int> idxVec2;
    
    for(string &g : goal)
    {
        if(cardsMap1.find(g) != cardsMap1.end())
        {
            idxVec1.push_back(cardsMap1[g]);
        }
        else if(cardsMap2.find(g) != cardsMap2.end())
        {
            idxVec2.push_back(cardsMap2[g]);
        }
    }
    
    string result = "Yes";
    for(int i=0; i<idxVec1.size(); i++)
    {
        if(idxVec1[i] != i)
        {
            result =  "No";
        }
    }
    for(int i=0; i<idxVec2.size(); i++)
    {
        if(idxVec2[i] != i)
        {
            result = "No";
        }
    }
    return result;
}